x-pack/elastic-agent/pkg/agent/application/emitter.go

```diff
@@ -30,54 +34,163 @@ type programsDispatcher interface {
 	Dispatch(id string, grpProg map[routingKey][]program.Program) error
 }
 
-func emitter(log *logger.Logger, router programsDispatcher, modifiers *configModifiers, reloadables ...reloadable) emitterFunc {
-	return func(c *config.Config) error {
-		if err := InjectAgentConfig(c); err != nil {
-			return err
+type emitterController struct {
+	logger      *logger.Logger
+	controller  composable.Controller
+	router      programsDispatcher
+	modifiers   *configModifiers
+	reloadables []reloadable
+
+	// state
+	lock   sync.RWMutex
+	config *config.Config
+	ast    *transpiler.AST
+	vars   []transpiler.Vars
+}
+
+func (e *emitterController) Update(c *config.Config) error {
+	if err := InjectAgentConfig(c); err != nil {
+		return err
+	}
+
+	// perform and verify ast translation
+	m, err := c.ToMapStr()
+	if err != nil {
+		return errors.New(err, "could not create the AST from the configuration", errors.TypeConfig)
+	}
+	rawAst, err := transpiler.NewAST(m)
+	if err != nil {
+		return errors.New(err, "could not create the AST from the configuration", errors.TypeConfig)
+	}
+	for _, filter := range e.modifiers.Filters {
+		if err := filter(e.logger, rawAst); err != nil {
+			return errors.New(err, "failed to filter configuration", errors.TypeConfig)
 		}
+	}
 
-		log.Debug("Transforming configuration into a tree")
-		m, err := c.ToMapStr()
+	// sanitary check that nothing in the config is wrong when it comes to variable syntax
+	ast := rawAst.Clone()
+	inputs, ok := transpiler.Lookup(ast, "inputs")
+	if ok {
+		renderedInputs, err := renderInputs(inputs, []transpiler.Vars{
+			{
+				Mapping: map[string]interface{}{},
+			},
+		})
 		if err != nil {
-			return errors.New(err, "could not create the AST from the configuration", errors.TypeConfig)
+			return err
 		}
-
-		ast, err := transpiler.NewAST(m)
+		err = transpiler.Insert(ast, renderedInputs, "inputs")
 		if err != nil {
-			return errors.New(err, "could not create the AST from the configuration", errors.TypeConfig)
+			return err
 		}
+	}
 
-		for _, filter := range modifiers.Filters {
-			if err := filter(log, ast); err != nil {
-				return errors.New(err, "failed to filter configuration", errors.TypeConfig)
+	programsToRun, err := program.Programs(ast)
+	if err != nil {
+		return err
+	}
+
+	for _, decorator := range e.modifiers.Decorators {
+		for outputType, ptr := range programsToRun {
+			programsToRun[outputType], err = decorator(outputType, ast, ptr)
+			if err != nil {
+				return err
 			}
 		}
+	}
+
+	e.lock.Lock()
```

Comment:
I keep stumbling over these lines and the ones on 112 and asking myself if we don't hit any race conditions somewhere or if we could do it in a nicer way.

Reply:
Being that its internal state to the struct that is called from 2 different go-routines, I don't really have another way of not requiring a lock.

This is also going to be important for the debugging commands to output the current state.